src/vm/instruction.ts

```diff
@@ -136,106 +136,75 @@ export function decodeInstruction(
-  const pcUpdateOff = UnsignedInteger.toUint16(7).unwrap();
+  const pcUpdateMask = 0x0380;
+  const pcUpdateOff = 7;
   // apUpdate is located at bits 10-11. We apply a mask of 0xc00 (0b110000000000)
   // and shift 10 bits right
-  const apUpdateMask = UnsignedInteger.toUint16(0x0c00).unwrap();
-  const apUpdateOff = UnsignedInteger.toUint16(10).unwrap();
+  const apUpdateMask = 0x0c00;
+  const apUpdateOff = 10;
   // opcode is located at bits 12-14. We apply a mask of 0x7000 (0b111000000000000)
   // and shift 12 bits right
-  const opcodeMask = UnsignedInteger.toUint16(0x7000).unwrap();
-  const opcodeOff = UnsignedInteger.toUint16(12).unwrap();
+  const opcodeMask = 0x7000;
+  const opcodeOff = 12;
 
   if ((highBit & encodedInstruction) !== 0n) {
     return new Err(HighBitSetError);
   }
 
   // mask for the 16 least significant bits of a 64-bit number
-  const mask = UnsignedInteger.toUint64(0xffffn).unwrap();
-  const shift = UnsignedInteger.toUint64(16n).unwrap();
+  const mask = 0xffffn;
+  const shift = 16n;
 
   // Get the offset by masking and shifting the encoded instruction
-  const offsetDstUint64 = UnsignedInteger.uint64And(encodedInstruction, mask);
-  const offsetDstResult = SignedInteger16.fromBiased(offsetDstUint64);
+  const offsetDstResult = SignedInteger16.fromBiased(encodedInstruction & mask);
   if (offsetDstResult.isErr()) {
     return offsetDstResult;
   }
   const offsetDst = offsetDstResult.unwrap();
 
-  let shiftedEncodedInstruction = UnsignedInteger.uint64Shr(
-    encodedInstruction,
-    shift
+  let shiftedEncodedInstruction = encodedInstruction >> shift;
+  const offsetOp0Result = SignedInteger16.fromBiased(
+    shiftedEncodedInstruction & mask
   );
-  const offsetOp0Uint64 = UnsignedInteger.uint64And(
-    shiftedEncodedInstruction,
-    mask
-  );
-  const offsetOp0Result = SignedInteger16.fromBiased(offsetOp0Uint64);
   if (offsetOp0Result.isErr()) {
     return offsetOp0Result;
   }
   const offsetOp0 = offsetOp0Result.unwrap();
 
-  shiftedEncodedInstruction = UnsignedInteger.uint64Shr(
-    shiftedEncodedInstruction,
-    shift
-  );
-  const offsetOp1Uint64 = UnsignedInteger.uint64And(
-    shiftedEncodedInstruction,
-    mask
+  shiftedEncodedInstruction = shiftedEncodedInstruction >> shift;
+  const offsetOp1Result = SignedInteger16.fromBiased(
+    shiftedEncodedInstruction & mask
   );
-  const offsetOp1Result = SignedInteger16.fromBiased(offsetOp1Uint64);
   if (offsetOp1Result.isErr()) {
     return offsetOp1Result;
   }
   const offsetOp1 = offsetOp1Result.unwrap();
 
   // Get the flags by shifting the encoded instruction
-  shiftedEncodedInstruction = UnsignedInteger.uint64Shr(
-    shiftedEncodedInstruction,
-    shift
-  );
-  const flagsResult = UnsignedInteger.downCastToUint16(
-    shiftedEncodedInstruction
-  );
-  if (flagsResult.isErr()) {
-    return flagsResult;
-  }
-
-  const flags = flagsResult.unwrap();
+  const flags = Number(shiftedEncodedInstruction >> shift);
```

Comment:
can this Number cast fail?
Why are we casting to a Number here

Reply:
nice naming, helped me understand where we are in the cursor of the whole encoded instruction